src/compiler/checker.ts

```diff
@@ -6113,11 +6113,12 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
         ) {
             const originalType = type;
             if (addUndefined) {
-                type = getOptionalType(type);
+                type = getOptionalType(type, !isParameter(host));
             }
             const clone = tryReuseExistingNonParameterTypeNode(context, typeNode, type, host);
             if (clone) {
-                if (addUndefined && !someType(getTypeFromTypeNode(context, typeNode), t => !!(t.flags & TypeFlags.Undefined))) {
+                // explicitly add `| undefined` if it's missing from the input type nodes and the type contains `undefined` (and not the missing type)
+                if (addUndefined && containsUndefinedType(type) && (type as UnionType).types[0] !== missingType && !someType(getTypeFromTypeNode(context, typeNode), t => !!(t.flags & TypeFlags.Undefined))) {

```

Comment:
Hm, rethinking, couldn't `type` itself be undefined in these cases? Is it safe to cast to `UnionType`?

Reply:
`containsUndefinedType` can only return `true` if `type` is a union, so the cast is safe.

Reply:
Wait, no, I'm mistaken, you're correct